database/database.go

```diff
@@ -1,127 +1,585 @@
+		Context varchar,
+		Text varchar,
+		Unixtime bigint,
+		Upvotes integer)`)
+	if err != nil {
+		postgresDatabase.Close()
+		return errors.New("At Setup: " + err.Error())
+	}
 
-	_, err := database.Exec("INSERT INTO quotes (teacherid, text) VALUES ($1, $2)", teacherid, quote)
+	// Create unverifiedQuotes table in PostgreSQL database if it doesn't exist
+	// for more information see UnverifiedQuoteT declaration
+	_, err = postgresDatabase.Exec(
+		`CREATE TABLE IF NOT EXISTS unverifiedQuotes (
+		QuoteID serial PRIMARY KEY,
+		Teacher varchar, 
+		Context varchar,
+		Text varchar,
+		Unictime bigint,
+		IPHash bigint)`)
 	if err != nil {
-		log.Print("Cannot store quote: ", err)
-		return err
+		postgresDatabase.Close()
+		return errors.New("At Setup: " + err.Error())
 	}
 
+	localDatabase.mux.Setup()
+	loadLocalDatabase()
+
 	return nil
+
 }
 
-// StoreTeacher is a function to store teachers
-func StoreTeacher(name string, title string, note string) error {
-	database.Ping()
+// Close database backend
+func Close() {
+	localDatabase.mux.LockWrite()
+	localDatabase.quoteSlice = nil
+	localDatabase.teacherSlice = nil
+	localDatabase.wordsMap = nil
+	localDatabase.mux.UnlockWrite()
+	postgresDatabase.Close()
+}
+
+/* -------------------------------------------------------------------------- */
+/*                          EXPORTED QUOTES FUNCTIONS                         */
+/* -------------------------------------------------------------------------- */
+
+// GetQuotes returns a slice containing all quotes
+// The weight variable will be zero
+func GetQuotes() *[]QuoteT {
+	localDatabase.mux.LockRead()
+	defer localDatabase.mux.UnlockRead()
+	quoteSlice := localDatabase.quoteSlice
+
+	return &quoteSlice
+}
+
+// GetQuotesByString returns a slice containing all quotes
+// The weight variable will indicate how well the given text matches the corresponding quote
+func GetQuotesByString(text string) *[]QuoteT {
+
+	localDatabase.mux.LockRead()
+	defer localDatabase.mux.UnlockRead()
+
+	quoteSlice := localDatabase.quoteSlice
 
-	_, err := database.Exec("INSERT INTO teachers (name, title, note) VALUES ($1, $2, $3)", name, title, note)
+	for word, count := range getWordsFromString(text) {
+		wordsMapItem := localDatabase.wordsMap[word]
+
+		_ = count
+		for _, v := range wordsMapItem.occurenceSlice {
+			quoteSlice[v.enumid].Match += float32(v.count) / float32(wordsMapItem.totalOccurences)
+		}
+	}
+	return &quoteSlice
+}
+
+// StoreQuote stores a new quote
+// If the ID is not zero, StoreQuote will try to find the appropriate quote and overwrite it
+// If the ID is zero a new quote will be created
+func StoreQuote(q QuoteT) error {
+	var err error
+
+	// Verify connection to PostgreSQL database
+	err = postgresDatabase.Ping()
 	if err != nil {
-		log.Print("Cannot store teacher: ", err)
-		return err
+		postgresDatabase.Close()
+		return errors.New("At StoreQuote: " + err.Error())
+	}
+
+	if q.QuoteID == 0 {
```

Comment:
please split the `StoreQuote` function into `CreateQuote` and `UpdateQuote`, especially because almost the entire function is an if/else statement. Splitting will have a few benefits:
- improves clarity of the code that's calling it
- shorter functions are easier to understand
- one function shall always be there for exactly one thing, and be named correspondingly
- makes execution path not dependent on a value in the slice (which would be difficult to track down)
- reduce bugs as a result